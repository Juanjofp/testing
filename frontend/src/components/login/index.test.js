import React from 'react';
import ReactDOM from 'react-dom';
import {shallow, mount, render} from 'enzyme';

import Login from './index';
import { wrap } from 'module';

const onClick = jest.fn();
const doLogin = jest.fn();
const clearErrorMessage = jest.fn();

window.__MUI_USE_NEXT_TYPOGRAPHY_VARIANTS__ = true;
jest.useFakeTimers();

describe(
    'Login component',
    () => {
        it(
            'renders without crashing', () => {
            const div = document.createElement('div');
            ReactDOM.render(
                <Login
                    location={{from: '/'}}/>,
                div);
            }
        );

        it(
            'create an snapshot',
            () => {
                const tree = render(
                    <Login location={{from: '/'}}/>
                );
                expect(tree).toMatchSnapshot();
            }
        );

        it(
            'create an snapshot while loading',
            () => {
                const tree = render(
                    <Login location={{from: '/'}} isLoading={true}/>
                );
                expect(tree).toMatchSnapshot();
            }
        );

        it(
            'create an snapshot with errors',
            () => {
                const tree = render(
                    <Login location={{from: '/'}} hasErrors={{message: 'Invalid user'}}/>
                );
                expect(tree).toMatchSnapshot();
            }
        );

        it(
            'renders login and password inputs',
            () => {
                const wraper = mount(
                    <Login location={{from: '/'}}/>
                );
                expect(wraper.find('[data-testid="login-email"]').exists()).toBe(true);
                expect(wraper.find('[data-testid="login-password"]').exists()).toBe(true);
                expect(wraper.find('[data-testid="login-submit"]').exists()).toBe(true);
            }
        );

        it(
            'renders button and submit credentials',
            () => {
                const wraper = mount(
                    <Login
                        location={{from: '/'}}
                        login={doLogin}
                        username='juanjo'
                        password='franco'/>,
                );
                const inputUsername = wraper.find('[data-testid="login-email"]').last();
                const inputPassword = wraper.find('[data-testid="login-password"]').last();
                const buttonLogin = wraper.find('[data-testid="login-submit"]').last();
                expect(inputUsername.exists()).toBe(true);
                expect(inputPassword.exists()).toBe(true);
                expect(buttonLogin.exists()).toBe(true);
                buttonLogin.props().onClick();
                expect(doLogin.mock.calls.length).toBe(1);
                expect(doLogin).toHaveBeenCalledWith('juanjo', 'franco');
            }
        );

        it(
            'renders errors and clear message after a few seconds',
            () => {
                const wraper = mount(
                    <Login
                        location={{from: '/'}}
                        hasErrors={{message: 'Invalid User'}}
                        loginClearError={clearErrorMessage}/>,
                );
                let errorText = wraper.find('[data-testid="error-message"]').first();
                expect(errorText.exists()).toBe(true);
                //console.log('>>>>>>>< Wrapper', errorText.debug());
                // Implementation details, never test this!!!
                expect(errorText.childAt(0).props().open).toBe(true);
                expect(clearErrorMessage.mock.calls.length).toBe(0);
                // Run all timers now!
                jest.runAllTimers();
                expect(clearErrorMessage.mock.calls.length).toBe(1);
            }
        );
    }
);