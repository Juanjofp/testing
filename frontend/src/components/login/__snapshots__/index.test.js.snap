// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login component create an snapshot 1`] = `
<div
  class="LoginPage"
>
  <div
    class="MuiPaper-root-2 MuiPaper-elevation1-5 MuiPaper-rounded-3 MuiCard-root-1"
  >
    <div
      class="LoginDialog"
    >
      <div
        class="AvatarLogin"
      >
        <div
          class="MuiAvatar-root-29"
          size="80"
        >
          <img
            alt="User default avatar"
            class="MuiAvatar-img-31"
            src="http://quizup-players.imgix.net/avatars/Avatar_astronaut-large-hd.png"
          />
        </div>
      </div>
      <div
        class="InputLogin"
      >
        <div
          class="MuiFormControl-root-32"
          data-testid="login-email"
        >
          <div
            class="MuiInputBase-root-49 MuiInput-root-36 MuiInput-underline-40 MuiInputBase-formControl-50 MuiInput-formControl-37"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input-59 MuiInput-input-44"
              id="login_username_field"
              type="text"
            />
          </div>
        </div>
      </div>
      <div
        class="InputLogin"
      >
        <div
          class="MuiFormControl-root-32"
          data-testid="login-password"
        >
          <div
            class="MuiInputBase-root-49 MuiInput-root-36 MuiInput-underline-40 MuiInputBase-formControl-50 MuiInput-formControl-37"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input-59 MuiInput-input-44 MuiInputBase-inputType-62 MuiInput-inputType-47"
              id="login_password_field"
              type="password"
            />
          </div>
        </div>
      </div>
      <div
        class="LoginButton"
      >
        <button
          class="MuiButtonBase-root-92 MuiButton-root-66 MuiButton-text-68 MuiButton-flat-71"
          data-testid="login-submit"
          id="login_submit_button"
          primary="LOGIN"
          style="width:240px"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label-67"
          >
            LOGIN
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Login component create an snapshot while loading 1`] = `
<div
  class="LoginPage"
>
  <div
    class="MuiPaper-root-2 MuiPaper-elevation1-5 MuiPaper-rounded-3 MuiCard-root-1"
  >
    <div
      class="LoginDialog"
    >
      <div
        class="AvatarLogin"
      >
        <div
          class="MuiAvatar-root-29"
          size="80"
        >
          <img
            alt="User default avatar"
            class="MuiAvatar-img-31"
            src="http://quizup-players.imgix.net/avatars/Avatar_astronaut-large-hd.png"
          />
        </div>
      </div>
      <div
        class="InputLogin"
      >
        <div
          class="MuiFormControl-root-32"
          data-testid="login-email"
        >
          <div
            class="MuiInputBase-root-49 MuiInput-root-36 MuiInput-underline-40 MuiInputBase-formControl-50 MuiInput-formControl-37"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input-59 MuiInput-input-44"
              id="login_username_field"
              type="text"
            />
          </div>
        </div>
      </div>
      <div
        class="InputLogin"
      >
        <div
          class="MuiFormControl-root-32"
          data-testid="login-password"
        >
          <div
            class="MuiInputBase-root-49 MuiInput-root-36 MuiInput-underline-40 MuiInputBase-formControl-50 MuiInput-formControl-37"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input-59 MuiInput-input-44 MuiInputBase-inputType-62 MuiInput-inputType-47"
              id="login_password_field"
              type="password"
            />
          </div>
        </div>
      </div>
      <div
        class="LoginButton"
      >
        <button
          class="MuiButtonBase-root-92 MuiButtonBase-disabled-93 MuiButton-root-66 MuiButton-text-68 MuiButton-flat-71 MuiButton-disabled-86"
          data-testid="login-submit"
          disabled=""
          id="login_submit_button"
          primary="LOGIN"
          style="width:240px"
          tabindex="-1"
          type="button"
        >
          <span
            class="MuiButton-label-67"
          >
            LOGIN
          </span>
        </button>
      </div>
      <span
        class="LoginLoading"
      >
        Loading...
      </span>
    </div>
  </div>
</div>
`;

exports[`Login component create an snapshot with errors 1`] = `
<div
  class="LoginPage"
>
  <div
    class="MuiPaper-root-2 MuiPaper-elevation1-5 MuiPaper-rounded-3 MuiCard-root-1"
  >
    <div
      class="LoginDialog"
    >
      <div
        class="AvatarLogin"
      >
        <div
          class="MuiAvatar-root-29"
          size="80"
        >
          <img
            alt="User default avatar"
            class="MuiAvatar-img-31"
            src="http://quizup-players.imgix.net/avatars/Avatar_astronaut-large-hd.png"
          />
        </div>
      </div>
      <div
        class="InputLogin"
      >
        <div
          class="MuiFormControl-root-32"
          data-testid="login-email"
        >
          <div
            class="MuiInputBase-root-49 MuiInput-root-36 MuiInput-underline-40 MuiInputBase-formControl-50 MuiInput-formControl-37"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input-59 MuiInput-input-44"
              id="login_username_field"
              type="text"
            />
          </div>
        </div>
      </div>
      <div
        class="InputLogin"
      >
        <div
          class="MuiFormControl-root-32"
          data-testid="login-password"
        >
          <div
            class="MuiInputBase-root-49 MuiInput-root-36 MuiInput-underline-40 MuiInputBase-formControl-50 MuiInput-formControl-37"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input-59 MuiInput-input-44 MuiInputBase-inputType-62 MuiInput-inputType-47"
              id="login_password_field"
              type="password"
            />
          </div>
        </div>
      </div>
      <div
        class="LoginButton"
      >
        <button
          class="MuiButtonBase-root-92 MuiButton-root-66 MuiButton-text-68 MuiButton-flat-71"
          data-testid="login-submit"
          id="login_submit_button"
          primary="LOGIN"
          style="width:240px"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label-67"
          >
            LOGIN
          </span>
        </button>
      </div>
      <div
        class="MuiSnackbar-root-95 MuiSnackbar-anchorOriginBottomCenter-97"
        data-testid="error-message"
      >
        <div
          class="MuiTypography-root-112 MuiTypography-body1-121 MuiPaper-root-2 MuiPaper-elevation6-10 MuiSnackbarContent-root-109"
          direction="up"
          role="alertdialog"
        >
          <div
            class="MuiSnackbarContent-message-110"
          >
            Invalid user
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
